Augment EditableTree (SharedTree) events

dstanesc opened the pull request:
## Description

Incremental delivery towards SharedTree data binding support. See complete [story](https://github.com/microsoft/FluidFramework/issues/14293).

Current enhancements
- Propagates `changing` event details
- Introduces additional `subtreeChanging` event

The new feature introduces no regressions to the existing functionality.

@CraigMacomber @DLehenbauer

dstanesc committed:
Augment editableTree events

CraigMacomber commented on packages/dds/tree/src/feature-libraries/editable-tree/editableTree.ts:
```diff
@@ -314,15 +314,35 @@ export class NodeProxyTarget extends ProxyTarget<Anchor> {
+						return (listener as EditableTreeEvents["subtreeChanging"])(anchorNode);
+					},
+				);
+				return () => {
+					unsubscribeFromSubtreeChange();
+				};
```
I think this would be slightly more efficient and simpler code.
Though if you want to keep it as is for consistency with the other code thats fine as well.
```suggestion
				return unsubscribeFromSubtreeChange;
```

dstanesc replied:
Looks good to me. Will perform the change as suggested.

CraigMacomber commented on packages/dds/tree/src/test/shared-tree/sharedTree.spec.ts:
```diff
@@ -598,28 +598,82 @@ describe("SharedTree", () => {
+
+		it("propagates path and value args for local and subtree changes", async () => {
+			const value = "42";
+			const provider = await TestTreeProvider.create(1);
```
Since this test does not need support for summarization, you can use the simpler and faster MockFluidDataStoreRuntime approach to making a test tree, see src/test/shared-tree/schmatizeTree.spec.ts for some examples.

That file also use the new schematizeView API, and typed schema apis, which would be preferred here if you are going to setup an actual shared-tree.

CraigMacomber replied:
I know a lot of our tests use the patterns you used here: we are trying to migrate away from them.

dstanesc replied:
Understood. Thanks for the pointers!

dstanesc replied:
Modified both event tests to use `MockFluidDataStoreRuntime` and `schematizeView` APIs.

CraigMacomber commented on packages/dds/tree/src/test/shared-tree/sharedTree.spec.ts:
```diff
@@ -598,28 +598,82 @@ describe("SharedTree", () => {
 			]);
 		});
+
+		it("propagates path and value args for local and subtree changes", async () => {
```
While its good to have some integration testing of the events, I think it would be preferred to do the detailed correctness testing in more targeted unit tests.

It should be possible to setup a build a forest with, and get an editable tree for that without involving the shared tree class or fluid at all. src/test/feature-libraries/editable-tree/editableTree.spec.ts has some examples of this.

I think manually triggering the events from the forest's anchorset (even if that requires casting its events object to an emitter) might be a cleaner way to unit test this (you don't actually have to edit the tree that way, and can just check the events subscribe and unsubscribe properly).

With just these two tests, I'm fine leaving it here as if you want, but if we are going to get significantly more of them in the future, I'd prefer more targeted unit tests.

dstanesc replied:
Added `editable-tree/editableTree.events.spec.ts` dedicated tests, thus resolving this conversation

CraigMacomber commented on packages/dds/tree/src/feature-libraries/editable-tree/editableTree.ts:
```diff
@@ -314,15 +314,35 @@ export class NodeProxyTarget extends ProxyTarget<Anchor> {
+				};
+			}
+			case "subtreeChanging": {
+				const unsubscribeFromSubtreeChange = this.anchorNode.on(
+					"subtreeChanging",
+					(anchorNode: AnchorNode) => {
+						return (listener as EditableTreeEvents["subtreeChanging"])(anchorNode);
+					},
```
I'd prefer to avoid this `as` cast if practical. I think you can avoid it by:
```suggestion
				const typedListener = listener;
				const unsubscribeFromSubtreeChange = this.anchorNode.on(
					"subtreeChanging",
					(anchorNode: AnchorNode) => typedListener(anchorNode)
```

or even:
I'd prefer to avoid this `as` cast if practical. I think you can avoid it by:
```suggestion
				const typedListener = listener;
				const unsubscribeFromSubtreeChange = this.anchorNode.on(
					"subtreeChanging",
					typedListener,
```

dstanesc replied:
Wonderful! I confirm, the additional declaration actually tweaks the compiler. Thank you for the insight!

dstanesc replied:
Actually appeared only to compile properly in the editor, probably some compilation lag... __Fails on the build.__ I am open to other suggestions as I also don't like the above trick. I know that following code also compiles:
```
const unsubscribeFromSubtreeChange = this.anchorNode.on(
  "subtreeChanging",
  (anchorNode: AnchorNode) => listener(anchorNode, undefined),
);
```
but the vscode hover call hints are off (ie misleading). Maybe Typescript version upgrade ?

dstanesc replied:
As unfortunately your suggestion did not work, eliminated the cast as outlined in my prior comment. Is this fine ?

CraigMacomber replied:
As I noted in a top level comment on this PR, this is fine, but the reasons are complicted.

dstanesc committed:
- Optimizes unsubscribeFromSubtreeChange return
- Removes as EditableTreeEvents["subtreeChanging"] cast
- Simplifies on("valueChanging", listener) call

dstanesc committed:
Editable-tree event subscription tests

dstanesc committed:
Correct build linting err

dstanesc committed:
optimize sharedTree event testing to use mockFluidDataStoreRuntime and schematizeView

dstanesc reviewed (commented):
Code ready for a 2nd review round when time permits. Performed all suggested updates except the cast workaround which actually did not work. Replaced the cast with a 2 argument invocation, if this is acceptable.

CraigMacomber commented:
> Code ready for a 2nd review round when time permits. Performed all suggested updates except the cast workaround which actually did not work. Replaced the cast with a 2 argument invocation, if this is acceptable.

TypeScript has some soundness issues related to optional function arguments (ex: https://github.com/microsoft/FluidFramework/blob/d4f08a3da698dc99a3ffed217e04fe15e1a7fe5a/docs/content/docs/concepts/Breaking%20Changes%20and%20Supported%20Patterns.md#optional-function-arguments ). As far as I can tell, explicitly providing extra undefined arguments should be ok (ex: https://www.typescriptlang.org/play?#code/LAKFDMFcDsGMBcCWB7aACcAKaAuN1IBbAIwFMAnNAXjQFYBKPAN2UQBM0BvUASFlQDOyADakAdMOQBzbPQDcoAL6gImGG1LhE0Um3lA) but it seems a bit messy and error prone (that approach won't keep working as our events get more augments of different types).

However, I figured out why my suggestion doesn't work: someone could call on like `tree.on<"changing" | "subtreeChanging">("subtreeChanging", (path: UpPath) =>{})`. Since this case is only an issue when the possible value the function can have overlap, the approach you took should actually continue to work in more complex cases, though it might also require adding an extra local variable in the future (or stop being necessary if the functions don't overlap).

Thus I think the approach you took here is fine, though I'm a bit annoyed with TypeScript since there is no way to prevent passing in unions of multiple keys :)
